### DISCUSSION TITLE
3.0.0

### DISCUSSION THREAD

**Discussion by mattesmohr** [vapor-community/HTMLKit#118] [General]
I have decided to remove the pre-rendering and the context management as it currently exists. As I think it was a good decision back then, now I think the swift language and the server performance has evolved since then, so it would be fair enough to drop it. 

Also I think a template engine makes more sense in cases, where the markup language and the language, the engine is written in, are two different languages. In HTMLKit, this ist not the case. 

I do know, dropping the pre-rendering costs performance, but I think the results are still acceptable. 

I think the decision is worth to bump the package version to 3.0.0. But of course there are more changes coming. I will list them here.

### Native statements

Dropping the pre-rendering allows us to use the native statements expression like `if`, `else`, `for` and string interpolation.

### View renderer

No pre-rendering means no cache. Therefore I had to change the Vapor provider again. Do use the provider you just have to call `request.htmlkit.render()` in your controller handler. Thats it. You don't have to add a view for pre-rendering anymore.

#### Alpha 1
https://github.com/vapor-community/HTMLKit/releases/tag/3.0.0-alpha.1

#### Alpha 2
https://github.com/vapor-community/HTMLKit/releases/tag/3.0.0-alpha.2

#### Alpha 3
https://github.com/vapor-community/HTMLKit/releases/tag/3.0.0-alpha.3

### Alpha 4
https://github.com/vapor-community/HTMLKit/releases/tag/3.0.0-alpha.4

### Alpha 5
https://github.com/vapor-community/HTMLKit/releases/tag/3.0.0-alpha.5

### Alpha 6
https://github.com/vapor-community/HTMLKit/releases/tag/3.0.0-alpha.4

**Comment by nonfungibletunji** [comment]
Where is the documentation for using?

**Reply by mattesmohr** [reply]
There is no documentation for 3.0.0 yet, as my focus right now, is more on creating a stable API. If things are more certain, I will update the documentation for sure. But you can ask me or you can have a look on https://github.com/mattesmohr/Website. It should be up to date to a certain point.

**Comment by nonfungibletunji** [comment]
Does the website actually work? What’s the URL, just curious….

**Reply by mattesmohr** [reply]
Nope, the website is currently offline. I understand. What are you interested in? Components, Performance?

**Comment by nonfungibletunji** [comment]
Be able to fully deploy a swift written website to vercel,

https://github.com/swift-cloud/Vercel

I have working code deployed to vercel using

https://github.com/BinaryBirds/swift-html

Working  and just trying to solidify my options….does the Svg tag work with HTMLKit

**Reply by mattesmohr** [reply]
I see. Well, I think it depends on what you like to do. I see HTMLKit as more than just a DSL. Also it's written only for Vapor at the moment. I cannot estimate what needs to be done to work with Vercel. In the near future there will be a preview page of HTMLKit. Maybe then you will get a better picture what HTMLKit is. So I would you recommend to wait and stay tuned.

**Comment by nonfungibletunji** [comment]
Any updates on documentation

**Reply by mattesmohr** [reply]
Hiya, you can find the recently added tutorials on [SPI](https://swiftpackageindex.com/vapor-community/htmlkit/main/tutorials/introduction) and the API documentation [also here](https://swiftpackageindex.com/vapor-community/HTMLKit/main/documentation/htmlkit).

Disclaimer: Still in progress
